File: ruby/test/task_master/task_repository_test.rb
require_relative "../test_helper"

class TaskMaster::TaskRepositoryTest < Minitest::Test
  def setup
    @mock_adapter = TestAdapter.new
    @class_name = TaskMaster::TaskRepository
  end

  def test_should_delegate_to_the_list_function_of_the_supplied_adapter
    expected = @mock_adapter.list.map { |x| x[:name] }
    assert_equal(expected, @class_name.new(@mock_adapter).list.map(&:name))
  end

  def test_should_filter_list_of_tasks_by_status
    expected = ["Foo"]
    assert_equal(expected, @class_name.new(@mock_adapter).list_by_status("Completed").map(&:name))
  end

  def test_should_filter_list_of_tasks_by_assignee
    expected = ["Baz"]
    assert_equal(expected, @class_name.new(@mock_adapter).list_by_assignee("John Smith").map(&:name))
  end

  def test_should_list_tasks_that_are_within_the_week_as_upcoming
    expected = ["Baz"]
    Date.stub :today, Date.parse("2023-11-12") do
      assert_equal(expected, @class_name.new(@mock_adapter).list_upcoming.map(&:name))
    end
  end

  def test_should_not_list_tasks_that_are_not_within_the_week_as_upcoming
    Date.stub :today, Date.parse("2023-11-12") do
      refute_includes(
        @class_name.new(@mock_adapter).list_upcoming.map(&:name),
        "Quux"
      )
    end
  end

  def test_should_list_tasks_as_overdue_if_they_happen_in_the_past_and_have_not_been_completed
    expected = ["Bar"]
    Date.stub :today, Date.parse("2023-11-12") do
      assert_equal(
        expected,
        @class_name.new(@mock_adapter).list_overdue.map(&:name)
      )
    end
  end

  def test_should_filter_list_of_tasks_by_ones_that_have_been_completed
    expected = ["Foo"]
    assert_equal(expected, @class_name.new(@mock_adapter).list_completed.map(&:name))
  end

  class TestAdapter
    def list
      [
        {
          id: "1",
          name: "Foo",
          assignee: "Jim Bob",
          due_on: "2023-11-11",
          status: "Completed"
        },
        {
          id: "2",
          name: "Bar",
          assignee: "Jim Bob",
          due_on: "2023-11-11",
          status: "In Progress"
        },
        {
          id: "3",
          name: "Baz",
          assignee: "John Smith",
          due_on: "2023-11-13",
          status: "Not Started"
        },
        {
          id: "4",
          name: "Quux",
          assignee: "Jane Doe",
          due_on: "2023-11-23",
          status: "Not Started"
        }
      ]
    end
  end
end

File: javascript/test/TaskRepository.test.js
const path = require('path');
const TaskRepository = require(path.join(__dirname, '../lib/TaskRepository'));

beforeEach(() => {
	mockAdapter = new TestAdapter();
	className = TaskRepository;
});

test("should delegate to the list function of the supplied adapter", () => {
	const expected = mockAdapter.list().map(element => {
		return element.name;
	});
	const actual = new className(mockAdapter).list().map(element => {
		return element.name;
	});
	expect(actual).toStrictEqual(expected);
});

test("should filter list of tasks by status", () => {
	const expected = ["Foo"];
	const actual = new className(mockAdapter).listByStatus({statuses: ["Completed"]}).map(element => {
		return element.name;
	});
	expect(actual).toStrictEqual(expected);
});

test("should filter list of tasks by assignee", () => {
	const expected = ["Baz"];
	const actual = new className(mockAdapter).listByAssignee({assignees: "John Smith"}).map(element => {
		return element.name;
	});
	expect(actual).toStrictEqual(expected);
});

test("should list tasks that are within the as upcoming", () => {
	const expected = ["Baz"];
	jest.useFakeTimers().setSystemTime(new Date('2023-11-12T00:00:00Z'));
	const actual = new className(mockAdapter).listUpcoming().map(element => {
		return element.name;
	});
	expect(actual).toStrictEqual(expected);
});

test("should list tasks are not within the week as upcoming", () => {
	jest.useFakeTimers().setSystemTime(new Date('2023-11-12T00:00:00Z'));
	const actual = new className(mockAdapter).listUpcoming().map(element => {
		return element.name;
	});
	expect(actual).not.toContain("Quux");
});

test("should list tasks as overdue if they happen in the past are incomplete", () => {
	const expected = ["Bar"];
	jest.useFakeTimers().setSystemTime(new Date('2023-11-12T00:00:00Z'));
	const actual = new className(mockAdapter).listOverdue().map(element => {
		return element.name;
	});
	expect(actual).toStrictEqual(expected);
});

test("should filter list of tasks by ones that have been completed", () => {
	const expected = ["Foo"];
	const actual = new className(mockAdapter).listComplete().map(element => {
		return element.name;
	});
	expect(actual).toStrictEqual(expected);
});


class TestAdapter {
	constructor() {}

	list() {
		return [
			{
				id: "1",
				name: "Foo",
				assignee: "Jim Bob",
				dueOn: "2023-11-11",
				status: "Completed"
			},
			{
				id: "2",
				name: "Bar",
				assignee: "Jim Bob",
				dueOn: "2023-11-11",
				status: "In Progress"
			},
			{
				id: "3",
				name: "Baz",
				assignee: "John Smith",
				dueOn: "2023-11-13",
				status: "Not Started"
			},
			{
				id: "4",
				name: "Quux",
				assignee: "Jane Doe",
				dueOn: "2023-11-23",
				status: "Not Started"
			}
		]
	}
}
